### DISCUSSION TITLE
Rate limit queue

### DISCUSSION THREAD

**Discussion by brycegoh** [riverqueue/river#133] [Ideas]
Hi, wondering if there is a way to rate limit a queue such as 10 jobs per minute, etc? Otherwise, is there a suggested alternative? Thanks!

**Comment by bgentry** [comment]
Hi @brycegoh, this isn’t something supported directly in River as of today, although it’s definitely something we’re thinking about and we have some ideas for how to make it happen. As of now I believe you’d have to layer on something within your jobs to make them hold until it’s their turn to run. This wouldn’t be a great option though as there would be actively running jobs taking up worker slots. I think this would have to come from the job system to work well.

I’ve converted this to a discussion so it can get upvotes alongside the other feature requests. Let us know if you think there’s anything particularly unique or interesting about your use case here. Thanks!

**Comment by adityathebe** [comment]
I assume for rate limits with such a small period `(like 10/minute)`, it's possible to use the [period based unique constraint](https://riverqueue.com/docs/unique-jobs#unique-by-period) and achieve a fairly accurate rate limiter?

For a larger window, a window based unique constraint would make it a feature complete rate limiter I suppose.

---
Disclaimer: I haven't actually used River. Saw this on Twitter yesterday & have only read about it.

**Comment by semanser** [comment]
I'm also interested in this feature. Any ideas on how to implement it?

**Comment by ligustah** [comment]
I have a somewhat related use case. Some of our jobs are talking to third party APIs and we want to (easily) limit the number of requests we are sending there way. Right now we do that by configuring the max workers for the queue directly, but this wouldn't allow us to scale the worker pool and even during deployments, when we are spinning up new workers, the parallel work is multiplied. So it would be very handy for us to set a global rate limit on the queues and not worry about the number of worker processes too much.

**Comment by AntiPaste** [comment]
Any news on this? I have a similar use case to a previous commenter where we are talking to an external API that imposes strict rate limits. Current solution of a separate queue for these rate limited jobs and checking / waiting for the rate limit to expire inside the running job feels very hacky.

**Comment by bgentry** [comment]
Somewhat related: we just launched a [concurrency limits](https://riverqueue.com/docs/pro/concurrency-limits) feature in River Pro. This allows you to set global limits on how many jobs run concurrently on a given queue, with optional partitioning.

While this is not the same as rate limiting, it may help solve some use cases. Real global rate limiting will be built on the same foundation and should hopefully follow in next couple months 🤞

**Comment by obs-gh-zuozhiwang** [comment]
Hi, a gentle bump on this topic, would really love to see this get implemented. I have a similar use case - rate limit on external API requests per key. I'm sending notifications to a lot of webhook callbacks, and I want to rate-limit (e.g. once per second) per webhook endpoint. 

I'd love to adopt riverqueue in our product and this is one of the blockers :)

**Comment by wbhob** [comment]
Same here – our specific use case is we are sending messages via Slack API which has a rate limit of 1 message per second, but we send large batches on a schedule for reminders to our clients. Having "1 job per second" for this specific job would make life much easier! 

For now our workaround will be using Job snoozing and a rate limiter at the app level since we can run everything on one instance right now, but at some point in the future once we're more distributed it would be nice to have this out of the box in River 🔥

**Comment by janblaha-arenko** [comment]
Also interested in rate limiting for a certain job - I am trying to schedule jobs with a minimum interval between all jobs of the same kind. 

Ideally, I would want to insert jobs and have any worker pick up the job (on any replica of the "service" for H-scaled systems) as soon as this "minimum interval" has passed (an this interval be guaranteed globally - across all replicas). 
Effectively - pop the jobs into the queue and pop a job every nth interval. When a job fails and can be retried, it goes back to the back of the queue.

**Reply by janblaha-arenko** [reply]
A slightly gnarly workaround I've thought of is to have a periodic job that runs every nth "interval" and picks up data required for the job (which is known ahead of time) from a separate store (DB/S3/you name it) which was inserted by separate job.

This effectively replicates that queue outside of River where each entry in the external store equates to a job that needs to be queued.

However, being able to do this natively in River would make this much smoother.
